File: scripts/pip-cache-dual-attention/op/torch_ops_extension/custom_ops/csrc/npu_dmp_sparse_flash_attention.cpp
/**
 * Copyright (c) 2025 Huawei Technologies Co., Ltd.
 * This program is free software, you can redistribute it and/or modify it under the terms and conditions of
 * CANN Open Software License Agreement Version 2.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 */

#include <iostream>
#include <torch/library.h>
#include "ops_common.h"

namespace custom {
using namespace at_npu::native;

// npu tensor max size
const int SIZE = 8;

// 工具函数，推导输出shape
at::Tensor construct_sparse_infer_output_tensor(
    const at::Tensor& query, const at::Tensor& value, std::string layout)
{
    for (size_t i = 0; i < query.sizes().size(); i++) {
        TORCH_CHECK(query.size(i) > 0, "All values within query's shape should be greater "
            "than 0, but shape[", i, "] is ", query.size(i));
    }
    at::Tensor output = at::empty(query.sizes(), query.options().dtype(query.dtype()));

    return output;
}

// step2, 为NPU设备实现前向接口
at::Tensor npu_dmp_sparse_flash_attention_npu(
    const at::Tensor &query, const at::Tensor &key, const at::Tensor &value,
    const at::Tensor &sparse_indices, double scale_value, int64_t sparse_block_size,
    const c10::optional<at::Tensor> &block_table,
    const c10::optional<at::Tensor> &actual_seq_lengths_query,
    const c10::optional<at::Tensor> &actual_seq_lengths_kv,
    const c10::optional<at::Tensor> &query_rope,
    const c10::optional<at::Tensor> &key_rope,
    const c10::optional<at::Tensor> &softmax_max_out,
    const c10::optional<at::Tensor> &softmax_sum_out,
    const c10::optional<at::Tensor> &prior_softmax_max,
    const c10::optional<at::Tensor> &prior_softmax_sum,
    const c10::optional<at::Tensor> &prior_attention_out,
    c10::string_view layout_query,
    c10::string_view layout_kv,
    int64_t sparse_mode)
{
    std::string layout_query_str = std::string(layout_query);
    std::string layout_kv_str = std::string(layout_kv);

    // construct the output tensor
    at::Tensor output = construct_sparse_infer_output_tensor(
        query, value, layout_query_str);
    // convert str
    char *layout_query_ptr = const_cast<char *>(layout_query_str.c_str());
    char *layout_kv_ptr = const_cast<char *>(layout_kv_str.c_str());

    EXEC_NPU_CMD_V1(aclnnDmpSparseFlashAttention, query,
        key, value, sparse_indices, block_table, actual_seq_lengths_query,
        actual_seq_lengths_kv, query_rope, key_rope, softmax_max_out, softmax_sum_out,
        prior_softmax_max, prior_softmax_sum, prior_attention_out, scale_value, sparse_block_size,
        layout_query_ptr, layout_kv_ptr, sparse_mode,
        output);
    return output;
}

// step3, 为META设备实现前向接口
at::Tensor npu_dmp_sparse_flash_attention_meta(
    const at::Tensor &query, const at::Tensor &key, const at::Tensor &value,
    const at::Tensor &sparse_indices, double scale_value, int64_t sparse_block_size,
    const c10::optional<at::Tensor> &block_table,
    const c10::optional<at::Tensor> &actual_seq_lengths_query,
    const c10::optional<at::Tensor> &actual_seq_lengths_kv,
    const c10::optional<at::Tensor> &query_rope,
    const c10::optional<at::Tensor> &key_rope,
    const c10::optional<at::Tensor> &softmax_max_out,
    const c10::optional<at::Tensor> &softmax_sum_out,
    const c10::optional<at::Tensor> &prior_softmax_max,
    const c10::optional<at::Tensor> &prior_softmax_sum,
    const c10::optional<at::Tensor> &prior_attention_out,
    c10::string_view layout_query,
    c10::string_view layout_kv,
    int64_t sparse_mode)
{
    std::string layout_query_str = std::string(layout_query);
    at::Tensor output = construct_sparse_infer_output_tensor(
        query, value, layout_query_str);

    return output;
}
}

// step4, 为NPU设备注册前向实现
TORCH_LIBRARY_IMPL(custom, PrivateUse1, m) {
    m.impl("npu_dmp_sparse_flash_attention", &custom::npu_dmp_sparse_flash_attention_npu);
}

// step5, 为META设备注册前向实现
TORCH_LIBRARY_IMPL(custom, Meta, m) {
    m.impl("npu_dmp_sparse_flash_attention", &custom::npu_dmp_sparse_flash_attention_meta);
}

File: scripts/pip-cache-dual-attention/op/torch_ops_extension/custom_ops/converter/npu_dmp_sparse_flash_attention.py
# Copyright (c) 2025 Huawei Technologies Co., Ltd.
# This program is free software, you can redistribute it and/or modify it under the terms and conditions of
# CANN Open Software License Agreement Version 2.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
# INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.

from typing import (
    Any, Callable, ContextManager, Iterator, List, Literal, NamedTuple, Optional, Sequence, Tuple, TypeVar,
    Union, overload,
)

import torch
import torchair
from torch import Generator, contiguous_format, inf, strided, SymInt
from torch.types import Device, Number, _bool, _complex, _device, _dtype, _float, _int, _layout, _qscheme, _size
from torchair._ge_concrete_graph import ge_apis as ge
from torchair._ge_concrete_graph.fx2ge_converter import declare_supported, register_fx_node_ge_converter
from torchair.ge._ge_graph import Tensor, TensorSpec, DataType
from torchair._ge_concrete_graph.supported_declaration import _TypedTensor, F32, F16, F64, I32, I16, I64, I8, U8, \
    BOOL, Support
from torchair.ge import attr


# 为自定义算子注册converter，用于torch.compile 场景成图
# 注意： meta_outputs形参名为固定写法，若写错会影响ge节点的输出dtype与shape推导
@register_fx_node_ge_converter(torch.ops.custom.npu_dmp_sparse_flash_attention.default)
def convert_npu_dmp_sparse_flash_attention(
    query: Tensor,
    key: Tensor,
    value: Tensor,
    sparse_indices: Tensor,
    scale_value: float,
    sparse_block_size: int,
    *,
    block_table: Optional[Tensor] = None,
    actual_seq_lengths_query: Optional[Tensor] = None,
    actual_seq_lengths_kv: Optional[Tensor] = None,
    query_rope: Optional[Tensor] = None,
    key_rope: Optional[Tensor] = None,
    softmax_max_out: Optional[Tensor] = None,
    softmax_sum_out: Optional[Tensor] = None,
    prior_softmax_max: Optional[Tensor] = None,
    prior_softmax_sum: Optional[Tensor] = None,
    prior_attention_out: Optional[Tensor] = None,
    layout_query: str = "BSND",
    layout_kv: str = "BSND",
    sparse_mode: int = 3,
    meta_outputs: TensorSpec = None,
):
    return torchair.ge.custom_op(
        "DmpSparseFlashAttention",
        inputs={"query": query,
                "key": key,
                "value": value,
                "sparse_indices": sparse_indices,
                "block_table": block_table,
                "actual_seq_lengths_query": actual_seq_lengths_query,
                "actual_seq_lengths_kv": actual_seq_lengths_kv,
                "query_rope": query_rope,
                "key_rope": key_rope,
                "softmax_max_out": softmax_max_out,
                "softmax_sum_out": softmax_sum_out,
                "prior_softmax_max": prior_softmax_max,
                "prior_softmax_sum": prior_softmax_sum,
                "prior_attention_out": prior_attention_out,
               },
        attrs={"scale_value": attr.Float(scale_value),
               "sparse_block_size": attr.Int(sparse_block_size),
               "layout_query": attr.Str(layout_query),
               "layout_kv": attr.Str(layout_kv),
               "sparse_mode": attr.Int(sparse_mode),
               },
        outputs=['attention_out']
    )
